docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+This appears to be of frankly little utility. Creating an aggregated signature
+incurs overhead, so frequently re-aggregating may incur a significant
+overhead. How costly this is is still subject to further investigation and
+performance testing.
+
+Each validator will still need to receive vote extension data
+from the peer validators in order for consensus to proceed. As a result, any
+advantage gained by aggregating signatures across the vote message will be
+nullified as a result of the addition of vote extensions.
+
+#### Block Creation
+
+When creating a block, the proposer may create a small set of short
+multi-signatures and attach these to the block instead of including one
+signature per validator.
+
+#### Block Verification
+
+Verification of blocks would not verify a set of many signatures. Verification
+would instead check the single multi-signature using the public keys stored
+by the validator. Currently, we verify each validator signature using
+the public key associated with that validator.
+
+#### IBC Relaying
+
+IBC would no longer need to transmit a large set of signatures when
+updating state. These state updates do not happen for every IBC packet, only
+when changing an IBC light client's view of the counterparty chain's state.
+General [packets][ibc-packet] only contain enough information to correctly
+route the data to the counterparty chain.
+
+IBC does persist commit signatures to the chain in these `MsgUpdateClient`
+message when updating state. This message would no longer need the full set
+of unique signatures and would instead only need one signature for all of the
+data in the header.
+
+Adding BLS signatures would create a new signature type that must be
+understood by the IBC module and by the relayers. For some operations, such
+as state updates, the set of data written into the chain and received by the
+IBC module would be slightly smaller.
+
+## Discussion
+
+### What are the proposed benefits to aggregated signatures?
+
+#### Reduce Block Size
+
+At the moment, the commit size contains a 512 bit signature for each validator
+that voted for the block. For the cosmos hub, which has 175 validators in the
+validator set, this amounts to about 11kb per block. That is around 113 GB over
+the lifetime of the chain's 10118561 blocks. (Note, the hub has increased the
+number of validators in the set over time so the signature size over the history
+of the chain is likely less than that value). 
+
+Signature aggregation would only produce two signatures for the entire block.
+One for the yays and one for the nays. Each BLS aggregated signature is 48
+bytes, per the [IETF standard of BLS signatures][bls-ietf-ecdsa-compare].
+Over the lifetime of the same cosmos hub chain, that would amount to about 1
+GB, a savings of 112 GB. While that is a large factor of reduction it's worth
+bearing in mind that, at [GCP's cost][gcp-storage-pricing] of $.026 USD per GB,
+that is a total savings of around $2.50 per month.
```

Comment:
```suggestion
Signature aggregation would only produce two signatures for the entire block:
One for the yeas and one for the nays. Each BLS aggregated signature is 48
bytes, per the [IETF standard of BLS signatures][bls-ietf-ecdsa-compare].
Over the lifetime of the same Cosmos Hub chain, that would amount to about 1
GiB, a savings of 112 GB. While that is a large factor of reduction it's worth
bearing in mind that, at [GCP's cost][gcp-storage-pricing] of $.026 USD per GB,
that is a total savings of around $2.50 per month.
```

Slightly careful here: I think GCS pricing is in base-10 GB not base-2 GiB, but the order is right. The other thing to consider is that "old" blockchain data can probably use lower-cost storage tiers, since they should presumably be WORM and frequency of reading old data should be relatively low.

(That said: Anyone storing blockchain data in a cloud should probably be asking some fundamental questions about how decentralized their network is, but that's not a concern for this doc)

The other factor to consider is how big the rest of the block is, and in particular what fraction of the block the signature data are. While 113 GiB seems like a lot, if a typical block is much larger than that (100 txn * 1 KiB, say), even that 11 KiB is only about 10% overhead. That might matter, but I think the ratio is more interesting than the raw number.

Reply:
> (That said: Anyone storing blockchain data in a cloud should probably be asking some fundamental questions about how decentralized their network is, but that's not a concern for this doc)

Fair enough. I was mostly trying to find a decent proxy for 'commodity cost of storing that amount of data' and figured that the GCP value was reasonable. I could certainly go pull up different values and investigate how much it would cost in cold storage vs. mounted volumes etc. but eh I don't think we're there _just_ yet. I think the point of this number is to give a ballpark but I'm happy to expand if it maybe isn't doing its office.

Reply:
> The other factor to consider is how big the rest of the block is, and in particular what fraction of the block the signature data are. While 113 GiB seems like a lot, if a typical block is much larger than that (100 txn * 1 KiB, say), even that 11 KiB is only about 10% overhead. That might matter, but I think the ratio is more interesting than the raw number.

Not a bad idea, if I get time I'll see if I can find a quick way to pull out a value for that.